tests/test-client-shape.lua

```diff
@@ -39,15 +41,142 @@ runner.run_steps{
         assert(not surface.load_silently(c.client_shape_clip, false))
         assert(not surface.load_silently(c.shape_clip, false))
 
+        assert(surface.load_silently(c.client_shape_input, false))
+        assert(surface.load_silently(c.shape_input, false))
+
         return true
     end,
 
     function()
         client.get()[1]:kill()
         return true
     end
+}
 
+local wb = nil
+local wb_ready = false
+local wb_clicks = 0
+table.insert(steps, function(count)
+    if count == 1 then
+        local widget = wibox.widget.base.make_widget()
+        function widget.draw()
+            wb_ready = true
+        end
+        local geometry = {
+            x = 0,
+            y = 0,
+            width = 500,
+            height = 500,
+        }
+        wb = wibox {
+            widget = widget,
+        }
+        wb:geometry(geometry)
+        wb.visible = true
+        wb.bg = "#888888FF"
+        wb:connect_signal("button::press", function(self, x, y, _, _)
+            print("Wibox click " .. x .. ", " .. y)
+            wb_clicks = wb_clicks + 1
+        end)
+    elseif wb_ready then
+        return true
+    end
+end)
 
-}
+local app_ready = false
+local app_exited = false
+local app_clicks = 0
+table.insert(steps, function(count)
+    if count == 1 then
+        spawn.with_line_callback(
+            "./test-client-shape-input 100 100 300 300",
+            {
+                stdout = function(line)
+                    print("Test app stdout: " .. line)
+
+                    if line == "Ready" then
+                        app_ready = true
+                    end
+                    if line:match("Click.*") then
+                        app_clicks = app_clicks + 1
+                    end
+                end,
+                stderr = function(line)
+                    print("Test app stderr: " .. line)
+                end,
+                exit = function(reason, code)
+                    print("Test app res: " .. code .. "(" .. reason .. ")")
+                    app_exited = true
+                end
+            })
+    elseif (app_ready) then
+        return true
+    end
+end)
+
+local function click(x, y, expect_hit)
+    table.insert(steps, function()
+        wb_clicks = 0
+        app_clicks = 0
+
+        -- DEBUG????????????????????
+        if expect_hit then
+            print("Waiting...")
+            os.execute("sleep 2")
+
+            -- local client = require("awful.client")
+            local c = client.get()[1]
+            client.focus = c
+            c:raise()
+            awesome.sync()
+
+            print("OK")
+        end
+        -- DEBUG????????????????????
+
+        mouse.coords{x=x, y=y}
+        assert(mouse.coords().x == x and mouse.coords().y == y)
+        root.fake_input("button_release", 1)
+        root.fake_input("button_press", 1)
```

Comment:
i think you may blindly copy the mouse-dragging logic,
as you first releasing the left button, and next keep it (indefinitely) pressed

Reply:
I don't follow. What mouse dragging logic? We want to click.